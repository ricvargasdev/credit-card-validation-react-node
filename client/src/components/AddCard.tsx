import React, { useState } from 'react';
import { CreditCard } from '../models/CreditCard';
import { Form, Button, Row, Col, Alert } from 'react-bootstrap';

interface Props {    
    addCard: (card: CreditCard) => void;
    errorMessage: string;
}

const AddCard = (props:Props) => {
    const [formData, setFormData] = useState({
        name: '',
        cardNumber: 0,
        limit: 0,
        balance: 0,
    });

    const [cardNumberError, setCardNumberError] = useState('');

    const handleChange = (event: { target: { name: any; value: any; }; }) => {
        const { name, value } = event.target;
        setFormData({ ...formData, [name]: value });
    };

    const handleSubmit = async (event: { preventDefault: () => void; }) => {
        event.preventDefault();

        if (formData.cardNumber.toString().length > 19) {
            setCardNumberError("Card number should be no longer than 19 characters");
            return;
        }
        setCardNumberError('');

        const addCard = await props.addCard({
            name: formData.name,
            cardNumber: formData.cardNumber,
            limit: formData.limit,
            balance: formData.balance,
        });

        setFormData({
            name: '',
            cardNumber: 0,
            limit: 0,
            balance: 0,
        });
    };

  return (
    <>
    <Row>
        <Col md={4}>
            <h2>Add</h2>
            <Form onSubmit={handleSubmit}>
                <Form.Group controlId="formName">
                    <Form.Label>Name</Form.Label>
                    <Form.Control 
                        type="text" 
                        placeholder="Enter name" 
                        name="name" 
                        value={formData.name} 
                        onChange={handleChange} 
                        required 
                        aria-label="Name input"
                    />
                </Form.Group>

                <Form.Group controlId="formCardNumber">
                    <Form.Label>Card Number</Form.Label>
                    <Form.Control 
                        type="number" 
                        placeholder="Enter card number" 
                        name="cardNumber" 
                        value={formData.cardNumber} 
                        onChange={handleChange} 
                        required
                        aria-label="Card number input"
                    />
                    {cardNumberError && <Alert key={'danger'} variant={'danger'}>{cardNumberError}</Alert>}
                    {props.errorMessage && <Alert key={'danger'} variant={'danger'}>{props.errorMessage}</Alert>}
                </Form.Group>

                <Form.Group controlId="formLimit">
                    <Form.Label>Limit</Form.Label>
                    <Form.Control 
                        type="number" 
                        placeholder="Enter limit" 
                        name="limit" 
                        value={formData.limit} 
                        onChange={handleChange} 
                        required
                        aria-label="Limit input"
                    />
                </Form.Group>

                <Form.Group controlId="formSubmit" style={{marginTop: 10}}>
                    <Button variant="primary" type="submit" aria-label="Add card button">
                        Add Card
                    </Button>
                </Form.Group>
            </Form>
        </Col>
    </Row>
    </>
  );
};

export default AddCard;
