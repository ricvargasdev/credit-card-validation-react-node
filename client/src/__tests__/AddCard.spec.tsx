import React from 'react';
import AddCard from '../components/AddCard';
import { create } from 'react-test-renderer';
import { render, fireEvent, act } from '@testing-library/react';

describe('Test for the AddCard component', () => {
  it('Renders the component', () => {
    const addCard = jest.fn();
    const component = create(<AddCard addCard={addCard} errorMessage={''} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('calls addCard function on form submit', () => {
    const addCard = jest.fn();
    const { getByLabelText, getByText } = render(<AddCard addCard={addCard} errorMessage={''} />);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'James Bond' } });
    fireEvent.change(getByLabelText('Card Number'), { target: { value: '1234567890' } });
    fireEvent.change(getByLabelText('Limit'), { target: { value: '100' } });
    fireEvent.click(getByText('Add Card'));

    expect(addCard).toHaveBeenCalledWith({ name: 'James Bond', cardNumber: '1234567890', limit: '100', balance: 0 });
  });
});