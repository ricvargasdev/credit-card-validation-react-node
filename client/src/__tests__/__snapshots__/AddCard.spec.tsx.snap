// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the AddCard component Renders the component 1`] = `
<div
  className="row"
>
  <div
    className="col-md-4"
  >
    <h2>
      Add
    </h2>
    <form
      className=""
      onSubmit={[Function]}
    >
      <div>
        <label
          className="form-label"
          htmlFor="formName"
        >
          Name
        </label>
        <input
          aria-label="Name input"
          className="form-control"
          id="formName"
          name="name"
          onChange={[Function]}
          placeholder="Enter name"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="formCardNumber"
        >
          Card Number
        </label>
        <input
          aria-label="Card number input"
          className="form-control"
          id="formCardNumber"
          name="cardNumber"
          onChange={[Function]}
          placeholder="Enter card number"
          required={true}
          type="number"
          value={0}
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="formLimit"
        >
          Limit
        </label>
        <input
          aria-label="Limit input"
          className="form-control"
          id="formLimit"
          name="limit"
          onChange={[Function]}
          placeholder="Enter limit"
          required={true}
          type="number"
          value={0}
        />
      </div>
      <div
        style={
          {
            "marginTop": 10,
          }
        }
      >
        <button
          aria-label="Add card button"
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Add Card
        </button>
      </div>
    </form>
  </div>
</div>
`;
